core/search/bm25.cpp

```diff
@@ -300,80 +300,187 @@ namespace bm25 {
+template<NormType Type, typename Reader>
 auto MakeNormAdapter(Reader&& reader) {
-  return NormAdapter<Reader, IsNorm2>(std::move(reader));
+  return NormAdapter<Reader, Type>(std::move(reader));
 }
 
+template<typename Ctx>
+struct MakeScoreFunctionImpl{
+  template<bool HasFilterBoost, typename... Args>
+  static score_function Make(Args&&... args);
+};
+
+template<>
+struct MakeScoreFunctionImpl<BM15Context> {
+  using Ctx = BM15Context;
+
+  template<bool HasFilterBoost, typename... Args>
+  static score_function Make(Args&&... args) {
+    return {
+        memory::make_unique<Ctx>(std::forward<Args>(args)...),
+        [](irs::score_ctx* ctx) noexcept -> const byte_type* {
+          auto& state = *static_cast<Ctx*>(ctx);
+
+          // FIXME???
+          const float_t tf = SQRT.get<true>(state.freq->value);
+
+          float_t c0;
+          if constexpr (HasFilterBoost) {
+            assert(state.filter_boost);
+            c0 = state.filter_boost->value * state.num;
+          } else {
+            c0 = state.num;
+          }
+
+          const float_t c1 = state.norm_const;
+
+          sort::score_cast<score_t>(state.score_buf) = c0 - c0 / (1.f + tf / c1);
+
+          return state.score_buf;
+        }
+    };
+  }
+};
+
 template<typename Norm>
-struct norm_score_ctx final : public score_ctx {
-  norm_score_ctx(
-      byte_type* score_buf,
-      float_t k, 
-      irs::boost_t boost,
-      const bm25::stats& stats,
-      const frequency* freq,
-      Norm&& norm,
-      const filter_boost* fb = nullptr) noexcept
-    : score_ctx{score_buf, k, boost, stats, freq, fb},
-      norm_{std::move(norm)} {
-    // if there is no norms, assume that b==0
-//    if (!norm_.empty()) {
-      norm_const_ = stats.norm_const;
-      norm_length_ = stats.norm_length;
- //   }
+struct MakeScoreFunctionImpl<BM25Context<Norm>> {
+  using Ctx = BM25Context<Norm>;
+
+  template<bool HasFilterBoost, typename... Args>
+  static score_function Make(Args&&... args) {
+    return {
+        memory::make_unique<Ctx>(std::forward<Args>(args)...),
+        [](irs::score_ctx* ctx) noexcept -> const byte_type* {
+          auto& state = *static_cast<Ctx*>(ctx);
+
+          float_t tf;
+          if constexpr (Norm::kType < NormType::kNorm) {
+            tf = static_cast<float_t>(state.freq->value);
+          } else {
+            tf = ::SQRT.get<true>(state.freq->value);
+          }
+
+          float_t c0;
+          if constexpr (HasFilterBoost) {
+            assert(state.filter_boost);
+            c0 = state.filter_boost->value * state.num;
+          } else {
+            c0 = state.num;
+          }
+
+          float_t c1;
+          if constexpr (NormType::kNorm2Tiny == Norm::kType) {
+            static_assert(std::is_same_v<uint32_t, decltype(state.norm())>);
+            c1 = state.norm_cache[state.norm() & 0xFF];
+          } else {
+            c1 = 1.f/(state.norm_const + state.norm_length * state.norm());
```

Comment:
Why can't we invert value in cache and replace `tf * c1` to `tf / c1` and then remove `1.f / `
Also, I think this
```C++
...(state.score_buf) = c0 - c0 / (1.f + tf / c1);
```
can be replaced by something like this
```C++
...(state.score_buf) = c0 - c0 * c1 / (c1 + tf);
```